lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3544,13 +3551,21 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     // Note: we relax the restriction if the type we're casting to is a
     // non-final class because it's possible that we might have a subclass
     // that conforms to the protocol.
+    //
+    //  Also, relax the restriction when toType is a stdlib collection
+    //  (Array, Set, Dictionary) that could have conditional conformances
+    //  because of custom casting mechanism implemented for those types e.g.
+    //
+    //  func encodable(_ value: Encodable) {
+    //    _ = value as! [String : Encodable]
+    //  }
```

Comment:
Could you move both of these notes (the relaxation of the check for non-final classes and the known stdlib collections) into the body of `couldDynamicallyConformToProtocol`?